### PR TITLE
Remove getting needless list length

### PR DESCRIPTION
**Pull request by auduchinok**
Removes some needless list traversals.

### CODE REVIEW

File: src/fsharp/absil/illib.fsi [hunk 1]
```diff
@@ -206,6 +206,8 @@ module internal List =
 
     val internal allEqual: xs:'T list -> bool when 'T: equality
 
+    val isSingleItem: xs: 'T list -> bool
```
Comment (dsyme) [inline src/fsharp/absil/illib.fsi:209]: I think `isSingleton` is more consistent (`List.isEmpty`, `List.singleton`) but no big deal

**Review by KevinRansom — APPROVED** [review]
I disagree with @dsyme here singleton has a particular meaning in programming. isSingleItem is a bit more specific here.

**Comment by vzarytovskii** [comment]
> I disagree with @dsyme here singleton has a particular meaning in programming. isSingleItem is a bit more specific here.

It also means "list with a single element" in many functional languages, besides, we already use it in lists and arrays `List.singleton`, `Array.singleton`.

**Comment by smoothdeveloper** [comment]
@KevinRansom, the contention is that `isSingleton` is almost always going to be `List.isSingleton`.

I think the reasoning of making it the pendant of `singleton` is good, in case we like to surface the function in FSharp.Core.

**Comment by auduchinok** [comment]
I've been actually choosing between `isSingleton` and `isSingleItem` names and decided to get with the latter. I'm OK with changing it to `isSingleton`, thanks!

**Review by vzarytovskii — APPROVED** [review]

**Comment by KevinRansom** [comment]
sounds good.

**Comment by smoothdeveloper** [comment]
@duckmatt / @knocte / @jgardella, wondering if this wouldn't be worth a rule in FSharpLint?

**Comment by knocte** [comment]
> wondering if this wouldn't be worth a rule in FSharpLint?

Good idea, although it would be much easier to make the violators of the rule respect it if there was an `isSingleton` function in FSharp's List, shouldn't we add that function first?

**Comment by smoothdeveloper** [comment]
@dsyme, a bit tangential, I remember some code reviews where we you preferred I change `singleton` calls into `[value]` literals, in the compiler code.

Do you feel `= [_]` could one day, be a legit operator application, and thus, a valid alternative to `isSingleton`?

It would be cool to have such bindings turn into boolean expression, if there is no ambiguity introduced in the language and such suggestion would makes sense.

**Comment by dsyme** [comment]
> you preferred I change singleton calls into [value] literals, in the compiler code.

Yes, `[expr]` is generally preferred.

> Do you feel `= [_]` could one day, be a legit operator application, and thus, a valid alternative to isSingleton?

The general construct would be "test if something matches a pattern and return a bool".   F# doesn't have this construct mainly out of simplicity, Probably `expr match? pat` or something would be the natural way of writing it.  But TBH I think F# is simpler without these numerous variations on pattern matching syntax cuteness, and it leads to less code churn between different ways of writing the same thing

**Comment by smoothdeveloper** [comment]
Thanks for the feedback, I think this, and https://github.com/fsprojects/FSharpLint/issues/543 would push us towards bringing the function in FSharp.Core, so FSharpLint can propose a quick fix to the issue (and embark the diagnostic about wrong code which I was concerned about).

**Pull request merged**